pyiron/atomistics/master/elastic.py

```diff
@@ -0,0 +1,385 @@
+        self.input['strain_matrices'] = []
+        self.input['use_symmetry'] = True
+        self.input['rotations'] = []
+        self.input['normalize_magnitude'] = False
+        self.input['use_elements'] = (True, 'whether or not consider chemical elements for '
+                                            + 'the symmetry analysis. Could be useful for SQS')
+        self.input['fit_first_order'] = False
+        self._job_generator = ElasticJobGenerator(self)
+
+    @property
+    def _number_of_measurements(self):
+        return int(max(
+            self.input['min_num_measurements'],
+            np.ceil(self.input['min_num_points']/max(len(self.input['rotations']), 1))
+        ))
+
+    def _get_rotation_matrices(self):
+        rotations = self.ref_job.structure.get_symmetry(use_elements=self.input['use_elements'])['rotations']
+        _, indices = np.unique(np.round(rotations, 6), axis=0, return_inverse=True)
+        rotations = rotations[np.unique(indices)]
+        self.input['rotations'] = rotations.tolist()
+
+    def _create_strain_matrices(self):
+        if self.input['use_symmetry'] and len(self.input['rotations'])==0:
+            self._get_rotation_matrices()
+        eps_lst = np.random.random((int(self._number_of_measurements), 3, 3))-0.5
+        eps_lst *= self.input['max_strain']
+        eps_lst += np.einsum('nij->nji', eps_lst)
+        self.input['strain_matrices'] = get_strain(max_strain=self.input['max_strain'],
+                                                   n_set=self._number_of_measurements,
+                                                   polynomial_order=self.input['polynomial_order'],
+                                                   additional_points=self.input['additional_points'],
+                                                   normalize=self.input['normalize_magnitude']).tolist()
+
+    def validate_ready_to_run(self):
+        super().validate_ready_to_run()
+        if len(self.input['strain_matrices'])==0:
+            self._create_strain_matrices()
+        if self.input['polynomial_order']<2:
+            raise ValueError('Minimum polynomial order: 2')
+        if self.input['polynomial_order']==2 and self.input['additional_points']>0:
+            warnings.warn('Setting additional points for anharmonic calculations only increases the number of calculations')
+        if self.input['polynomial_order']==2 and self.input['normalize_magnitude']:
+            warnings.warn('Magnitude normalization could reduce accuracy in harmonic calculations')
+        if self.input['polynomial_order']>2 and not self.input['normalize_magnitude']:
+            warnings.warn('Not normalizing magnitude could destabilise fit procedure')
+        if self.input['fit_first_order']:
+            warnings.warn('first order fit does not correct the second order coefficients in the current implementation')
+
+    def collect_output(self):
+        if self.ref_job.server.run_mode.interactive:
```

Comment:
Nit: This `if/else` looks like it needs a refactor. I'm just writing in the comment and not testing it so take this with a grain of salt, but I think the following should be equivalent:

```python
for key in ['energy_tot', 'energy_pot', 'pressures', 'volume']:
    if self.ref_job.server.run_mode.interactive:
        self.output[key] = self._read_child_output(self.project_hdf5.inspect(self.child_ids[0]))
    else:
        self.output[key] = [self._read_child_output(self.project_hdf5.inspect(job_id)) for job_id in self.child_ids]
```

where

```python
def _read_child_output(hdf, key):
    if key in hdf["output/generic"].list_nodes():
        return hdf["output/generic/{}".format(key)][-1])
    else:
        return []
```

This cuts 19 lines down to 10 and (IMO) improves readability.

Reply:
Note: I also sneakily use `self.output` instead of `self._output` so I have an excuse to point it out and remind you that `__init__` should have `self._output = InputList(table_name='output')` and there should be an `output` property and setter wrapping assignments as `InputList` ;) 

(You might need to fiddle with the table name...for `input` there's definitely no conflict, for output I'm not as sure.)

Reply:
Actually I copied it from `Murnaghan` without thinking much about it, but I guess the reason why it's done this way is because it performs one `inspect` for each job, and probably that makes the parsing faster.

Reply:
> Note: I also sneakily use `self.output` instead of `self._output` so I have an excuse to point it out and remind you that `__init__` should have `self._output = InputList(table_name='output')` and there should be an `output` property and setter wrapping assignments as `InputList` ;)
> 
> (You might need to fiddle with the table name...for `input` there's definitely no conflict, for output I'm not as sure.)

And I still don't know how to use `InputList` properly XD. Do you think if I simply add `self._output = InputList(table_name='output')` it's gonna start working straightaway?

Reply:
> Actually I copied it from `Murnaghan` without thinking much about it, but I guess the reason why it's done this way is because it performs one `inspect` for each job, and probably that makes the parsing faster.

Ah, good point. I still find the code messy though. What about something like this:

```python
if self.ref_job.server.run_mode.interactive:
    ids = [self.child_ids[0]]
else:
    ids = [job_id for job_id in self.child_ids]
self._output = self._add_energy_pressure_volume_output(self._output, ids)
```

```python
def _add_energy_pressure_volume_output(self, output, ids):
    for id_ in ids:
        hdf = self.project_hdf5.inspect(id_)
        for key in ['energy_tot', 'energy_pot', 'pressures', 'volume']:
            val = _read_output_key_or_empty(hdf, key)
            try:
                output[key].append(val)
            except KeyError:
                output[key] = [val]
    # I'm not sure about the formatting here -- 
    #    I'm wrapping the interactive case as a list, and never converting the regular case to np.array
    #    But there must be *some* way to do this nicely without as much duplicate code as we have now

@staticmethod
def _read_output_key_or_empty(hdf, key):
    # I suspect this is useful all over the codebase atm
    if key in hdf["output/generic"].list_nodes():
        return hdf["output/generic/{}".format(key)][-1])
    else:
        return []
```

Reply:
> And I still don't know how to use `InputList` properly XD. Do you think if I simply add `self._output = InputList(table_name='output')` it's gonna start working straightaway?

Yes, should do. You'll need to add `self._output.to_hdf(self._hdf5)` and `self._output.from_hdf(self._hdf5)` (or extremely similar?) in the `to_` and `from_hdf` methods. Other than that you should be good to go.